File: dotnet/MockResponses.cs
namespace CardPaymentSample;

/// <summary>
/// Mock responses for testing payment scenarios
/// </summary>
public static class MockResponses
{
    /// <summary>
    /// Get successful payment response
    /// </summary>
    public static PaymentResponse GetPaymentResponse(decimal amount, string paymentMethodId)
    {
        var random = new Random();
        
        return new PaymentResponse
        {
            TransactionId = $"txn_{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}_{Guid.NewGuid().ToString()[^9..]}",
            Amount = amount,
            Currency = "USD",
            Status = "approved",
            ResponseCode = "00",
            ResponseMessage = "Approved",
            Timestamp = DateTime.UtcNow,
            GatewayResponse = new GatewayResponse
            {
                AuthCode = $"A{random.Next(10000, 99999):D5}",
                ReferenceNumber = $"REF{random.Next(100000000, 999999999):D10}"
            }
        };
    }


    /// <summary>
    /// Get decline response with specific reason
    /// </summary>
    public static DeclineResponse GetDeclineResponse(string reason)
    {
        var declineReasons = new Dictionary<string, string>
        {
            {"insufficient_funds", "Insufficient Funds"},
            {"generic", "Card Declined"},
            {"pickup_card", "Pick Up Card"},
            {"lost_card", "Lost Card"},
            {"stolen_card", "Stolen Card"},
            {"expired_card", "Expired Card"},
            {"incorrect_cvc", "Incorrect CVC"},
            {"incorrect_zip", "Incorrect ZIP"},
            {"card_declined", "Card Declined"},
            {"invalid_account", "Invalid Account"},
            {"card_not_activated", "Card Not Activated"},
            {"processing_error", "Processing Error"},
            {"system_error", "System Error"}
        };

        var message = declineReasons.TryGetValue(reason, out var msg) ? msg : "Card Declined";

        return new DeclineResponse
        {
            ErrorCode = reason.ToUpper(),
            ResponseMessage = message
        };
    }

    /// <summary>
    /// Generate mock vault token
    /// </summary>
    public static string GenerateMockVaultToken()
    {
        return $"token_{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}_{Guid.NewGuid().ToString().Replace("-", "")[..16]}";
    }
}

File: php/MockResponses.php
<?php

declare(strict_types=1);

/**
 * Mock responses for testing without SDK
 */
class MockResponses
{
    /**
     * Generate mock stored payment token response
     */
    public static function getStoredPaymentToken(array $cardData): array
    {
        return [
            'id' => 'stored_' . uniqid() . '_' . bin2hex(random_bytes(8)),
            'brand' => $cardData['brand'],
            'last4' => $cardData['last4'],
            'exp_month' => $cardData['exp_month'],
            'exp_year' => $cardData['exp_year'],
            'created_at' => date('c'),
            'type' => 'card'
        ];
    }

    /**
     * Get card details from mock stored payment token
     */
    public static function getCardDetailsFromToken(string $storedPaymentToken): array
    {
        // Extract mock data from token pattern or use defaults for demo
        $mockDetails = [
            'brand' => 'Visa',
            'last4' => '0016',
            'expiryMonth' => '12',
            'expiryYear' => '28'
        ];

        // If token contains identifiable patterns, use them
        if (strpos($storedPaymentToken, 'visa') !== false) {
            $mockDetails['brand'] = 'Visa';
            $mockDetails['last4'] = '0016';
        } elseif (strpos($storedPaymentToken, 'mastercard') !== false || strpos($storedPaymentToken, 'mc') !== false) {
            $mockDetails['brand'] = 'Mastercard';
            $mockDetails['last4'] = '5780';
        } elseif (strpos($storedPaymentToken, 'amex') !== false) {
            $mockDetails['brand'] = 'American Express';
            $mockDetails['last4'] = '1018';
        } elseif (strpos($storedPaymentToken, 'discover') !== false) {
            $mockDetails['brand'] = 'Discover';
            $mockDetails['last4'] = '6527';
        }

        return $mockDetails;
    }

    /**
     * Generate mock payment response
     */
    public static function getPaymentResponse(float $amount, string $paymentMethodId): array
    {
        return [
            'transaction_id' => 'txn_' . uniqid(),
            'amount' => $amount,
            'currency' => 'USD',
            'status' => 'approved',
            'response_code' => '00',
            'response_message' => 'Approved',
            'timestamp' => date('c'),
            'payment_method_id' => $paymentMethodId,
            'gateway_response' => [
                'auth_code' => strtoupper(bin2hex(random_bytes(3))),
                'reference_number' => 'ref_' . uniqid()
            ]
        ];
    }



    /**
     * Generate decline responses for testing
     */
    public static function getDeclineResponse(string $reason): array
    {
        $responses = [
            'insufficient_funds' => [
                'response_code' => '51',
                'response_message' => 'Insufficient funds',
                'error_code' => 'CARD_DECLINED'
            ],
            'expired_card' => [
                'response_code' => '54',
                'response_message' => 'Expired card',
                'error_code' => 'EXPIRED_CARD'
            ],
            'invalid_card' => [
                'response_code' => '14',
                'response_message' => 'Invalid card number',
                'error_code' => 'INVALID_CARD'
            ]
        ];

        return $responses[$reason] ?? [
            'response_code' => '05',
            'response_message' => 'Do not honor',
            'error_code' => 'GENERIC_DECLINE'
        ];
    }
}